pkg/skaffold/build/gcb/status.go

```diff
@@ -0,0 +1,197 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package gcb
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strings"
+	"sync"
+	"time"
+
+	"google.golang.org/api/cloudbuild/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
+
+	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/gcp"
+)
+
+var (
+	reporter     *statusReporterImpl
+	reporterOnce sync.Once
+	client       *cloudbuild.Service
+	clientOnce   sync.Once
+)
+
+type statusReporter interface {
+	getStatus(ctx context.Context, projectID string, buildID string) <-chan result
+}
+
+type statusReporterImpl struct {
+	results         map[jobID]chan result
+	resultMutex     sync.RWMutex
+	requests        chan jobRequest
+	pollingInterval time.Duration
+}
+
+type jobID struct {
+	projectID string
+	buildID   string
+}
+
+type jobRequest struct {
+	jobID
+	ctx context.Context
+}
+
+type result struct {
+	jobID
+	status *cloudbuild.Build
+	err    error
+}
+
+func getStatusReporter() statusReporter {
+	reporterOnce.Do(func() {
+		reporter = &statusReporterImpl{
+			pollingInterval: PollingInterval,
+			results:         make(map[jobID]chan result),
+			requests:        make(chan jobRequest, 100),
+		}
+
+		go reporter.start()
+	})
+	return reporter
+}
+
+func (r *statusReporterImpl) start() {
+	poll := time.NewTicker(r.pollingInterval)
+	jobsByProjectID := make(map[string]map[jobID]jobRequest)
+	for {
+		select {
+		case req := <-r.requests:
+			if _, found := jobsByProjectID[req.projectID]; !found {
+				jobsByProjectID[req.projectID] = make(map[jobID]jobRequest)
+			}
+			jobsByProjectID[req.projectID][req.jobID] = req
+		case <-poll.C:
+			for projectID, jobs := range jobsByProjectID {
+				for id, req := range jobs {
+					if req.ctx.Err() != nil {
+						r.setResult(id, result{jobID: id, err: req.ctx.Err()})
+						delete(jobs, id)
+					}
+				}
+				statuses, err := getStatuses(projectID, jobs)
+				if err != nil {
+					for id := range jobs {
+						r.setResult(id, result{jobID: id, err: err})
+						delete(jobs, id)
+					}
+					continue
+				}
+				for id := range jobs {
+					cb := statuses[id]
+					switch cb.Status {
+					case StatusQueued, StatusWorking, StatusUnknown:
+					case StatusSuccess:
+						r.setResult(id, result{jobID: id, status: cb})
+						delete(jobs, id)
+					case StatusFailure, StatusInternalError, StatusTimeout, StatusCancelled:
+						r.setResult(id, result{jobID: id, err: fmt.Errorf("cloud build failed: %s", cb.Status)})
+						delete(jobs, id)
+					default:
+						r.setResult(id, result{jobID: id, err: fmt.Errorf("unknown status: %s", cb.Status)})
```

Comment:
Is this really different from the `StatusUnknown` case?  Maybe `unhandled status` might be more to point.

Reply:
changed